File: client/src/modules/nexotype/schemas/engineering/design-mutation.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const DesignMutationSchema = z.object({
  id: z.number(),
  candidate_id: z.number().int(),
  position: z.number().int(),
  wild_type: z.string().min(1).max(10),
  mutant: z.string().min(1).max(10),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateDesignMutationSchema = z.object({
  candidate_id: z.number().int(),
  position: z.number().int(),
  wild_type: z.string().min(1).max(10),
  mutant: z.string().min(1).max(10),
});

export const UpdateDesignMutationSchema = z.object({
  candidate_id: z.number().int().optional(),
  position: z.number().int().optional(),
  wild_type: z.string().min(1).max(10).optional(),
  mutant: z.string().min(1).max(10).optional(),
});

// ==========================================
// Types
// ==========================================

export type DesignMutation = z.infer<typeof DesignMutationSchema>;
export type CreateDesignMutation = z.infer<typeof CreateDesignMutationSchema>;
export type UpdateDesignMutation = z.infer<typeof UpdateDesignMutationSchema>;

// ==========================================
// Response Types
// ==========================================

export type DesignMutationResponse = {
  success: boolean;
  data?: DesignMutation;
  error?: string;
};

export type DesignMutationsResponse = {
  success: boolean;
  data?: DesignMutation[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/engineering/design_mutation_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class DesignMutationCreate(BaseModel):
    """Schema for creating a new design mutation"""
    candidate_id: int
    position: int
    wild_type: str
    mutant: str


class DesignMutationUpdate(BaseModel):
    """Schema for updating a design mutation"""
    candidate_id: int = None
    position: int = None
    wild_type: str = None
    mutant: str = None


# ==========================================
# Response Schemas
# ==========================================

class DesignMutationDetail(BaseModel):
    """Schema for design mutation details"""
    id: int
    candidate_id: int
    position: int
    wild_type: str
    mutant: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class DesignMutationListResponse(BaseModel):
    """Response schema for listing design mutations"""
    success: bool
    data: list[DesignMutationDetail] | None = None
    count: int = 0
    error: str | None = None


class DesignMutationResponse(BaseModel):
    """Response schema for single design mutation operations"""
    success: bool
    data: DesignMutationDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
